src/gainmap.c

```diff
@@ -286,4 +299,327 @@ avifResult avifImageApplyGainMap(const avifImage * baseImage,
+        totalCount += histogram[i];
+    }
+
+    const int maxOutliersOnEachSide = (int)avifRoundf(totalCount * MAX_OUTLIERS_RATIO / 2.0f);
+
+    int leftOutliers = 0;
+    *rangeMin = BUCKET_MIN_VALUE;
+    for (int i = 0; i < NUM_HISTOGRAM_BUCKETS; ++i) {
+        leftOutliers += histogram[i];
+        if (leftOutliers > maxOutliersOnEachSide) {
+            break;
+        }
+        if (histogram[i] == 0) {
+            *rangeMin = avifBucketIdxToValue(i);
+        }
+    }
+
+    int rightOutliers = 0;
+    *rangeMax = BUCKET_MAX_VALUE;
+    for (int i = NUM_HISTOGRAM_BUCKETS - 1; i >= 0; --i) {
+        rightOutliers += histogram[i];
+        if (rightOutliers > maxOutliersOnEachSide) {
+            break;
+        }
+        if (histogram[i] == 0) {
+            *rangeMax = avifBucketIdxToValue(i);
+        }
+    }
+}
+
+avifResult avifComputeGainMapRGB(const avifRGBImage * baseRgbImage,
+                                 avifTransferCharacteristics baseTransferCharacteristics,
+                                 const avifRGBImage * altRgbImage,
+                                 avifTransferCharacteristics altTransferCharacteristics,
+                                 avifColorPrimaries colorPrimaries,
+                                 avifGainMap * gainMap,
+                                 avifDiagnostics * diag)
+{
+    avifDiagnosticsClearError(diag);
+
+    AVIF_CHECKERR(baseRgbImage != NULL && altRgbImage != NULL && gainMap != NULL && gainMap->image != NULL, AVIF_RESULT_INVALID_ARGUMENT);
+    if (baseRgbImage->width != altRgbImage->width || baseRgbImage->height != altRgbImage->height) {
+        avifDiagnosticsPrintf(diag, "Both images should have the same dimensions");
+        return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+    if (gainMap->image->width == 0 || gainMap->image->height == 0 || gainMap->image->depth == 0 ||
+        gainMap->image->yuvFormat <= AVIF_PIXEL_FORMAT_NONE || gainMap->image->yuvFormat >= AVIF_PIXEL_FORMAT_COUNT) {
+        avifDiagnosticsPrintf(diag, "gianMap->image should be non null with desired width, height, depth and yuvFormat set");
+        return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+
+    const int width = baseRgbImage->width;
+    const int height = baseRgbImage->height;
+
+    avifRGBColorSpaceInfo baseRGBInfo;
+    avifRGBColorSpaceInfo altRGBInfo;
+    if (!avifGetRGBColorSpaceInfo(baseRgbImage, &baseRGBInfo) || !avifGetRGBColorSpaceInfo(altRgbImage, &altRGBInfo)) {
+        avifDiagnosticsPrintf(diag, "Unsupported RGB color space");
+        return AVIF_RESULT_NOT_IMPLEMENTED;
+    }
+
+    float * gainMapF[3] = { 0 }; // Temporary buffers for the gain map as floating point values, one per RGB channel.
+    avifRGBImage gainMapRGB;
+    memset(&gainMapRGB, 0, sizeof(gainMapRGB));
+    avifImage * gainMapImage = gainMap->image;
+
+    avifResult res = AVIF_RESULT_OK;
+    // --- After this point, the function should exit with 'goto cleanup' to free allocated resources.
```

Comment:
I know it's widespread but I thought it was useful to document when we start allocating resources and must use goto? Before this point, the input checks can do simple returns.
Not sure using a helper function is that useful? All avifComputeGainMap() below does it allocate some resources then call a helper function, and I still thought the goto was useful (as much as I hate gotos, but that's C for you).

Reply:
> I know it's widespread but I thought it was useful to document when we start allocating resources and must use goto? Before this point, the input checks can do simple returns.

Fine with me.

> Not sure using a helper function is that useful?

I was not talking about a helper function, but to split the function into:

```c
avifResult avifFunctionImpl(avifImage ** image) {
    *image = avifImageCreateEmpty();
    AVIF_CHECKERR(*image != NULL, AVIF_RESULT_OOM);
    AVIF_CHECKRES(avifImageAllocate(*image));
    ...
    return AVIF_RESULT_OK;
}
avifResult avifFunction() {
    avifImage * image = NULL;
    const avifResult result = avifFunctionImpl(&image);
    if (image != NULL) {
        avifImageDestroy(image);
    }
    return result;
}
```
but this is a personal preference, `goto`s work too.

Reply:
I see. I don't think this solution is great either because now we have to free resources allocated by a different function. Splitting alloc and free into different functions is not intuitive and just as error prone in my opinion. But again, memory management in C is inherently error prone...